### PR TITLE
reserve time for each action in login thread

### LINKED ISSUE #5172: 5.14 Error: script ran too long on first login
### Is there an existing issue for this?

- [X] I have searched the existing open and closed issues.

### Description

Going from 5.13.2 to 5.14 gives me a Script ran for too long and none of the weakauras show, typing /wa does nothing either, a quick /reloadUI fixes it and the error don't come back.

seems to be only the first time after the update, my weakauras.lua is rather big (33mb) attached below

thanks in advance

### WeakAuras Version

WeakAuras 5.14

### World of Warcraft Flavor

Cataclysm Classic

### World of Warcraft Region

US/NA

### Tested with only WeakAuras

I got this issue with only WeakAuras enabled

### Lua Error

```Text
1x WeakAuras/WeakAuras.lua:1299: script ran too long
[string "@WeakAuras/WeakAuras.lua"]:1299: in function `Login'
[string "@WeakAuras/WeakAuras.lua"]:1386: in function <WeakAuras/WeakAuras.lua:1334>

Locals:
initialTime = 15000
takeNewSnapshots = true
loginThread = <no value>
startTime = 161794208.926200
finishTime = 161805358.110600
ok = true
msg = nil
(*temporary) = true
(*temporary) = nil
(*temporary) = true
(*temporary) = true
(*temporary) = nil
(*temporary) = "script ran too long"
Private = <table> {
 HandleGlowAction = <function> defined @WeakAuras/WeakAuras.lua:3616
 frame_strata_types = <table> {
 }
 DisplayToString = <function> defined @WeakAuras/Transmission.lua:357
 combat_event_type = <table> {
 }
 regions = <table> {
 }
 frameLevels = <table> {
 }
 CheckItemSlotCooldowns = <function> defined @WeakAuras/GenericTrigger.lua:3166
 InitializeEncounterAndZoneLists = <function> defined @WeakAuras/Types_Cata.lua:13
 IsEnvironmentInitialized = <function> defined @WeakAuras/AuraEnvironment.lua:200
 form_types = <table> {
 }
 CleanArchive = <function> defined @WeakAuras/History.lua:25
 UpdateProgressFrom = <function> defined @WeakAuras/RegionTypes/RegionPrototype.lua:567
 RunConditions = <function> defined @WeakAuras/Conditions.lua:840
 tooltip_count = <table> {
 }
 SmoothStatusBarMixin = <table> {
 }
 pet_spec_types = <table> {
 }
 regionOptions = <table> {
 }
 get_zoneId_list = <function> defined @WeakAuras/Types_Cata.lua:75
 checkForSingleLoadCondition = <function> defined @WeakAuras/Prototypes.lua:1224
 author_option_classes = <table> {
 }
 grid_types = <table> {
 }
 non_transmissable_fields = <table> {
 }
 absorb_modes = <table> {
 }
 combatlog_spell_school_types_for_ui = <table> {
 }
 miss_types = <table> {
 }
 CancelDelayedTrigger = <function> defined @WeakAuras/GenericTrigger.lua:944
 centered_types_h = <table> {
 }
 ensurePRDFrame = <function> defined @WeakAuras/WeakAuras.lua:5434
 talent_types = <table> {
 }
 LoadFunction = <function> defined @WeakAuras/AuraEnvironment.lua:652
 reset_swing_spells = <table> {
 }
 spec_types_all = <table> {
 }
 player_target_events = <table> {
 }
 FinishLoadUnload = <function> defined @WeakAuras/WeakAuras.lua:2064
 subRegionTypes = <table> {
 }
 UnregisterAllEveryFrameUpdate = <function> defined @WeakAuras/GenericTrigger.lua:1884
 array_entry_name_types = <table> {
 }
 eclipse_direction_types = <table> {
 }
 StringToTable = <function> defined @WeakAuras/Transmission.lua:301
 RegisterLoadEvents = <function> defined @WeakAuras/WeakAuras.lua:1910
 GetReputationsSorted = <function> defined @WeakAuras/Types.lua:1712
 blend_types = <table> {
 }
 text_automatic_width = <table> {
 }
 EnforceSubregionExists = <function> defined @WeakAuras/RegionTypes/RegionPrototype.lua:1125
 CheckSpellCooldown = <function> defined @WeakAuras/GenericTrigger.lua:3022
 custom_trigger_types = <table> {
 }
 group_types = <table> {
 }
 subRegionOptions = <table> {
 }
 GetProgressValueConstant = <function> defined @WeakAuras/WeakAuras.lua:3927
 text_rotate_types = <table> {
 }
 spec_types = <table> {
 }
 anim_ease_types = <table> {
 }
 item_slot_types = <table> {
 }
 multiUnitUnits = <table> {
 }
 anim_color_types = <table> {
 }
 loaded = <table> {
 }
 Convert = <function> defined @WeakAuras/WeakAuras.lua:2281
 sound_channel_types = <table> {
 }
 CheckCooldownReady = <function> defined @WeakAuras/GenericTrigger.lua:3224
 event_prototypes = <table> {
 }
 NeedToRepairDatabase = <function> defined @WeakAuras/WeakAuras.lua:2358
 AtlasList = <table> {
 }
 RegisterRegionType = <function> defined @WeakAuras/WeakAuras.lua:413
 classification_types = <table> {
 }
 GetCurrencyListSize = <function> defined =[C]:-1
 unit_types_bufftrigger_2 = <table> {
 }
 IsOptionsProcessingPaused = <function> defined @WeakAuras/WeakAuras.lua:1586
 ValueFromPath = <function> defined @WeakA
```

### Reproduction Steps

1. Update addon
2. Login to the game
3. ???
4. get error

### Last Good Version

Weakauras 5.13.2

### Screenshots

Saved Variables before updating.
[WeakAuras.zip](https://github.com/user-attachments/files/15920651/WeakAuras.zip)

### Export String

_No response_

### Bisector Report

_No response_

### PR DESCRIPTION
**Pull request by emptyrivers**
the times selected are completely arbitrary. The primary reason for this is to make WA liable to defer Resume() until the next frame, if AddMany() takes a decent chunk of time (usually due to migrations).

A user who's really determined can still defeat this strategy by importing so many auras that Resume() itself takes > 19 seconds to execute, but there's very little we can do about that anyways.

Closes #5172 (not by fixing it, but by making it less likely)

### CODE REVIEW

**Comment by Desik** [comment]
I tested the experimental build and I keep getting the same error

`1x WeakAuras/WeakAuras.lua:1301: script ran too long
[string "@WeakAuras/WeakAuras.lua"]:1301: in function `Login'
[string "@WeakAuras/WeakAuras.lua"]:1389: in function <WeakAuras/WeakAuras.lua:1337>

Locals:
initialTime = 15000
takeNewSnapshots = true
loginThread = <no value>
startTime = 240853857.499900
finishTime = 240864908.433600
ok = true
val = nil
timeForNextStep = 1000
(*temporary) = true
(*temporary) = nil
(*temporary) = true
(*temporary) = true
(*temporary) = nil
(*temporary) = "script ran too long"
Private = <table> {
 HandleGlowAction = <function> defined @WeakAuras/WeakAuras.lua:3619
 frame_strata_types = <table> {
 }
 DisplayToString = <function> defined @WeakAuras/Transmission.lua:357
 combat_event_type = <table> {
 }
 regions = <table> {
 }
 frameLevels = <table> {
 }
 CheckItemSlotCooldowns = <function> defined @WeakAuras/GenericTrigger.lua:3166
 InitializeEncounterAndZoneLists = <function> defined @WeakAuras/Types_Cata.lua:13
 IsEnvironmentInitialized = <function> defined @WeakAuras/AuraEnvironment.lua:200
 form_types = <table> {
 }
 CleanArchive = <function> defined @WeakAuras/History.lua:25
 UpdateProgressFrom = <function> defined @WeakAuras/RegionTypes/RegionPrototype.lua:567
 RunConditions = <function> defined @WeakAuras/Conditions.lua:840
 tooltip_count = <table> {
 }
 SmoothStatusBarMixin = <table> {
 }
 pet_spec_types = <table> {
 }
 regionOptions = <table> {
 }
 get_zoneId_list = <function> defined @WeakAuras/Types_Cata.lua:75
 checkForSingleLoadCondition = <function> defined @WeakAuras/Prototypes.lua:1224
 author_option_classes = <table> {
 }
 grid_types = <table> {
 }
 non_transmissable_fields = <table> {
 }
 absorb_modes = <table> {
 }
 combatlog_spell_school_types_for_ui = <table> {
 }
 miss_types = <table> {
 }
 CancelDelayedTrigger = <function> defined @WeakAuras/GenericTrigger.lua:944
 centered_types_h = <table> {
 }
 ensurePRDFrame = <function> defined @WeakAuras/WeakAuras.lua:5437
 talent_types = <table> {
 }
 LoadFunction = <function> defined @WeakAuras/AuraEnvironment.lua:652
 reset_swing_spells = <table> {
 }
 spec_types_all = <table> {
 }
 player_target_events = <table> {
 }
 FinishLoadUnload = <function> defined @WeakAuras/WeakAuras.lua:2067
 subRegionTypes = <table> {
 }
 UnregisterAllEveryFrameUpdate = <function> defined @WeakAuras/GenericTrigger.lua:1884
 array_entry_name_types = <table> {
 }
 eclipse_direction_types = <table> {
 }
 StringToTable = <function> defined @WeakAuras/Transmission.lua:301
 RegisterLoadEvents = <function> defined @WeakAuras/WeakAuras.lua:1913
 GetReputationsSorted = <function> defined @WeakAuras/Types.lua:1712
 blend_types = <table> {
 }
 text_automatic_width = <table> {
 }
 EnforceSubregionExists = <function> defined @WeakAuras/RegionTypes/RegionPrototype.lua:1125
 CheckSpellCooldown = <function> defined @WeakAuras/GenericTrigger.lua:3022
 custom_trigger_types = <table> {
 }
 group_types = <table> {
 }
 subRegionOptions = <table> {
 }
 GetProgressValueConstant = <function> defined @WeakAuras/WeakAuras.lua:3930
 text_rotate_types = <table> {
 }
 spec_types = <table> {
 }
 anim_ease_types = <table> {
 }
 item_slot_types = <table> {
 }
 multiUnitUnits = <table> {
 }
 anim_color_types = <table> {
 }
 loaded = <table> {
 }
 Convert = <function> defined @WeakAuras/WeakAuras.lua:2284
 sound_channel_types = <table> {
 }
 CheckCooldownReady = <function> defined @WeakAuras/GenericTrigger.lua:3224
 event_prototypes = <table> {
 }
 NeedToRepairDatabase = <function> defined @WeakAuras/WeakAuras.lua:2361
 AtlasList = <table> {
 }
 RegisterRegionType = <function> defined @WeakAuras/WeakAuras.lua:413
 classification_types = <table> {
 }
 GetCurrencyListSize = <function> defined =[C]:-1
 unit_types_bufftrigger_2 = <table> {
 }
 IsOptionsProcessingPaused = <function> defined @WeakAuras/WeakAuras.lua:1589
 ValueFromPath = <function> defined `

File: WeakAuras/WeakAuras.lua [hunk 1]
```diff
@@ -1382,7 +1371,7 @@ loadedFrame:SetScript("OnEvent", function(self, event, addon)
       dbIsValid = true
     end
     if dbIsValid then
-      -- run login thread for up to 15 seconds, then defer to dynFrame
+      -- run login thread for up to 15 seconds, then defer to Threads
```
Comment (InfusOnWoW) [inline WeakAuras/WeakAuras.lua:1374]: the comment here is now wrong

File: WeakAuras/WeakAuras.lua [hunk 2]
```diff
@@ -4305,78 +4294,130 @@ function WeakAuras.EnsureString(input)
...
-    if dynFrame.update[name] then
-      dynFrame.update[name] = nil;
-      dynFrame.size = dynFrame.size - 1
-      if dynFrame.size == 0 then
-        dynFrame.frame:Hide();
-      end
+  ---@param name string
+  ---@param prio threadPriority
+  function threads:SetPriority(name, prio)
+    if self.prios[name] then
```
Comment (InfusOnWoW) [inline WeakAuras/WeakAuras.lua:4335]: calling with an unchanged priority seems to not work, e.g. calling with "normal" while self.prios[name] is alreaedy normal will run:

self["normal"][name] = self["normal"][name]
self["normal"][name] = nil

**Review by InfusOnWoW — APPROVED** [review]

File: WeakAuras/WeakAuras.lua (hunk 1 above)
Comment (emptyrivers) [reply WeakAuras/WeakAuras.lua:1374]: I think it's been wrong for a long time i.e. the 15000 we pass into Login is used as a boolean value.

**Comment by Desik** [comment]
I tested the last artifact and got an error again

`1x WeakAuras/WeakAuras.lua:4377: script ran too long
[string "@WeakAuras/WeakAuras.lua"]:4377: in function <WeakAuras/WeakAuras.lua:4363>
[string "@WeakAuras/WeakAuras.lua"]:4397: in function `Immediate'
[string "@WeakAuras/WeakAuras.lua"]:1296: in function `Login'
[string "@WeakAuras/WeakAuras.lua"]:1370: in function <WeakAuras/WeakAuras.lua:1318>

Locals:
pool = <table> {
 login = <no value>
}
finish = 276505524.720800
defaultEstimate = 1000
start = 276490524.721600
estimates = <table> {
 login = 5
}
ok = true
val = 5
continue = true
(for generator) = <function> defined =[C]:-1
(for state) = <table> {
 login = <no value>
}
(for control) = "login"
name = "login"
thread = <no value>
estimate = 5
(*temporary) = true
(*temporary) = 5
(*temporary) = true
(*temporary) = 5
(*temporary) = true
(*temporary) = "script ran too long"
debugprofilestop = <function> defined =[C]:-1
debugstack = <function> defined =[C]:-1
threads = <table> {
 normal = <table> {
 }
 SetPriority = <function> defined @WeakAuras/WeakAuras.lua:4338
 instant = <table> {
 }
 urgent = <table> {
 }
 Remove = <function> defined @WeakAuras/WeakAuras.lua:4348
 Immediate = <function> defined @WeakAuras/WeakAuras.lua:4395
 Add = <function> defined @WeakAuras/WeakAuras.lua:4324
 background = <table> {
 }
 frame = Frame {
 }
 prios = <table> {
 }
 size = 1
}`

File: WeakAuras/WeakAuras.lua [hunk 3]
```diff
@@ -2615,7 +2599,7 @@ function Private.AddMany(tbl, takeSnapshots)
         oldSnapshots[data.uid] = Private.GetMigrationSnapshot(data.uid)
       end
       Private.SetMigrationSnapshot(data.uid, data)
-      coroutine.yield()
+      coroutine.yield(500)
```
Comment (InfusOnWoW) [inline WeakAuras/WeakAuras.lua:2602]: This might be too low

**Comment by emptyrivers** [comment]
whoops, dont mind me - this comment will get [deleted](https://www.youtube.com&#x2F;watch?v=dQw4w9WgXcQ) in about 5 minutes

**Review by InfusOnWoW — APPROVED** [review]

**Pull request merged**